openedx/core/djangoapps/user_api/preferences/api.py

```diff
@@ -0,0 +1,198 @@
+    if preference_value is None:
+        raise PreferenceNotFound()
+    return preference_value
+
+
+def get_user_preferences(requesting_user, username=None):
+    """Returns all user preferences as a JSON response.
+
+    Args:
+        requesting_user (User): The user requesting the user preferences. Only the user with username
+            `username` or users with "is_staff" privileges can access the preferences.
+        username (str): Optional username for which to look up the preferences. If not specified,
+            `requesting_user.username` is assumed.
+
+    Returns:
+         A dict containing account fields.
+
+    Raises:
+         AccountUserNotFound: no user with username `username` exists (or `requesting_user.username` if
+            `username` is not specified)
+        AccountNotAuthorized: the requesting_user does not have access to the user preference.
+    """
+    existing_user = _get_user(requesting_user, username, allow_staff=True)
+    user_serializer = UserSerializer(existing_user)
+    return user_serializer.data["preferences"]
+
+
+def update_user_preferences(requesting_user, update, username=None):
+    """Update the user preferences for the given username.
+
+    Note:
+        It is up to the caller of this method to enforce the contract that this method is only called
+        with the user who made the request.
+
+    Arguments:
+        requesting_user (User): The user requesting to modify account information. Only the user with username
+            'username' has permissions to modify account information.
+        update (dict): The updated account field values. Note that null values for a preference will
+            be treated as a request to delete the key in question.
+        username (string): Optional username specifying which account should be updated. If not specified,
+            `requesting_user.username` is assumed.
+
+    Raises:
+        AccountUserNotFound: no user with username `username` exists (or `requesting_user.username` if
+            `username` is not specified)
+        AccountNotAuthorized: the requesting_user does not have access to change the account
+            associated with `username`
+        PreferenceUpdateError: the update could not be completed.
+    """
+    existing_user = _get_user(requesting_user, username)
+    try:
+        # First validate each preference setting
+        errors = {}
+        for preference_key in update.keys():
+            preference_value = update[preference_key]
+            try:
+                if preference_value is not None:
+                    UserPreference.validate_preference(existing_user, preference_key, preference_value)
+            except ValidationError as error:
+                errors[preference_key] = {
+                    "developer_message": error.message
+                }
+        if errors:
+            raise PreferenceValidationError(errors)
```

Comment:
Note that this raise will be caught by the outer try. You need to split up the two try clauses.